### PR TITLE
use oncecell on w, zkpm and srs

### PR DESCRIPTION
**Pull request by chee-chyuan**

### CODE REVIEW

**Review by mimoo — DISMISSED** [review]
Beautiful! There's a number of manual deserialization of `zk_w3` on the mina side. I can take a look shortly.

**Comment by mimoo** [comment]
looks like this is still a draft, so won't merge now

**Comment by chee-chyuan** [comment]
> looks like this is still a draft, so won't merge now

yea. i also plan to do the same to other parameters which inly depends on verifier_index. let me do that before i mark it as ready

**Comment by chee-chyuan** [comment]
i also changed abit on how srs is being loaded `from_file` and added an error for that. lmk if you need i need to amend it

File: kimchi/src/verifier_index.rs
Comment (mrmr1993) [inline kimchi/src/verifier_index.rs:243]: I don't think we want to create a new SRS here, for memory usage / performance reasons. Can we continue to use the one from the prover index?

File: kimchi/src/verifier_index.rs
Comment (mrmr1993) [inline kimchi/src/verifier_index.rs:69]: This field (and the others below) need to be public, otherwise the mina side won't be able to create a verifier index from one it receives over the network.

**Review by mrmr1993 — CHANGES_REQUESTED** [review]

File: kimchi/src/verifier_index.rs
Comment (chee-chyuan) [reply kimchi/src/verifier_index.rs:69]: my bad on this. not aware how mina does things

File: kimchi/src/verifier_index.rs
Comment (mrmr1993) [inline kimchi/src/verifier_index.rs:182]: If you change this to
```rust
srs: {
    let cell = OnceCell::new();
    cell.set(Arc::clone(&self.srs));
    cell
},
```
then we get back the old behaviour while still allowing for deserialisation.

Comment (chee-chyuan) [reply kimchi/src/verifier_index.rs:182]: arr ok i gotcha!

File: kimchi/src/verifier_index.rs
Comment (mimoo) [reply kimchi/src/verifier_index.rs:244]: I think you should change the field to be `Option<Arc<SRS>>`, like that it will deserialize to `None`. You can then implement `srs()` as:

```rust
fn srs(&self) -> Arc<SRS> {
  self.srs.expect("You must use `set_srs` to initialize the SRS")
}

  fn set_srs(&mut self, srs: Arc<SRS>) {
    self.srs = Some(srs)
  }
```

Comment (chee-chyuan) [reply kimchi/src/verifier_index.rs:244]: but that loses the ‘auto’ deserialization property though. or we dont really need that for srs?

Comment (mimoo) [reply kimchi/src/verifier_index.rs:244]: Oh no you’re right. I was thinking that we’d never want to do this, but actually we’d probably want to do it. We can probably use max_poly_size in the verifier index to init it if it’s not initialized

File: kimchi/src/error.rs
Comment (mimoo) [inline kimchi/src/error.rs:72]: maybe say something like "SRS already set"

**Comment by mimoo** [comment]
this looks good to me now! But @mrmr1993 I'll let you approve

**Review by mrmr1993 — APPROVED** [review]
Looks great, thank you!

**Pull request merged**
